napps/kytos/of_core/main.py

```diff
@@ -108,29 +109,82 @@ def handle_features_reply(self, event):
 
         switch.update_features(features)
 
-    @listen_to('kytos/core.messages.openflow.new')
-    def handle_new_openflow_message(self, event):
+        if (event.source.state == CONNECTION_STATE.SETUP and
+                event.source.protocol.state == 'waiting_features_reply'):
+            event.source.protocol.state = 'handshake_complete'
+            event.source.state = CONNECTION_STATE.ESTABLISHED
+            log.info('Handshake complete on connection: %s', event.source.id)
+            event_raw = KytosEvent(name='kytos/of_core.handshake_complete',
+                                   content={'source': event.source})
+            self.controller.buffers.app.put(event_raw)
+
+    @listen_to('kytos/core.openflow.data.in')
+    def handle_data_in(self, event):
         """Handle a RawEvent and generate a kytos/core.messages.in.* event.
 
         Args:
             event (KytosEvent): RawEvent with openflow message to be unpacked
         """
-        log.debug('RawOpenFlowMessage received by RawOFMessage handler')
-
-        # creates an empty OpenFlow Message based on the message_type defined
-        # on the unpacked header.
-        message = new_message_from_header(event.content['header'])
-        binary_data = event.content['binary_data']
-
-        # The unpack will happen only to those messages with body beyond header
-        if binary_data and len(binary_data) > 0:
-            message.unpack(binary_data)
-        log.debug('RawOpenFlowMessage unpacked')
 
+        connection = event.source
+        if connection.state in (CONNECTION_STATE.FINISHED,
+                                CONNECTION_STATE.FAILED):
+            return
+
+        data = connection.remaining_data + event.content['new_data']
+        packets, connection.remaining_data = of_slicer(data)
+        if not packets:
+            return
+
+        unprocessed_packets = []
+
+        for packet in packets:
+            log.debug('New Raw Openflow packet')
+            log.debug(packet)
+
+            message = unpack(packet)
+            if not message:
```

Comment:
This if may be replaced with a "try/except" if `unpack` raises an error instead of returning False.